File: backend/specs/akvo/lumen/specs/aggregation.clj
(ns akvo.lumen.specs.aggregation
  (:require [akvo.lumen.lib.aggregation :as aggregation]
            [akvo.lumen.lib.aggregation.bar :as aggregation.bar]
            [akvo.lumen.lib.aggregation.bubble :as aggregation.bubble]
            [akvo.lumen.lib.aggregation.line :as aggregation.line]
            [akvo.lumen.lib.aggregation.pie :as aggregation.pie]
            [akvo.lumen.lib.aggregation.pivot :as aggregation.pivot]
            [akvo.lumen.lib.aggregation.scatter :as aggregation.scatter]
            [akvo.lumen.postgres.filter :as postgres.filter]
            [akvo.lumen.specs.db :as db.s]
            [akvo.lumen.specs.db.dataset-version :as db.dsv.s]
            [akvo.lumen.specs.db.dataset-version.column :as db.dsv.column.s]
            [clojure.spec.alpha :as s]))

(s/def ::aggregation/dataset (s/keys :req-un [::db.dsv.s/columns ::db.dsv.s/table-name]))
(s/def ::aggregation.pie/bucketColumn ::db.dsv.column.s/columnName)
(s/def ::aggregation.pie/query (s/keys :req-un [::postgres.filter/filters
                                                ::aggregation.pie/bucketColumn]))

(s/fdef aggregation.pie/query
  :args (s/cat
         :db-conn ::db.s/tenant-connection
	 :dataset ::aggregation/dataset
	 :query ::aggregation.pie/query)
  :ret any?)

(s/def ::aggregation.pivot/categoryColumn (s/nilable ::db.dsv.column.s/columnName)) ;; todo should be fixed in client side?
(s/def ::aggregation.pivot/rowColumn (s/nilable ::db.dsv.column.s/columnName)) ;; todo should be fixed in client side?
(s/def ::aggregation.pivot/valueColumn (s/nilable ::db.dsv.column.s/columnName)) ;; todo should be fixed in client side?
(s/def ::aggregation.pivot/aggregation #{"mean"
                                         "sum"
                                         "min"
                                         "max"
                                         "count"})

(s/def ::aggregation.pivot/query (s/keys :req-un [::postgres.filter/filters
                                                  ::aggregation.pivot/aggregation]
                                         :opt-un [::aggregation.pivot/rowColumn
                                                  ::aggregation.pivot/valueColumn
                                                  ::aggregation.pivot/categoryColumn]))

(s/fdef aggregation.pivot/query
  :args (s/cat
         :db-conn ::db.s/tenant-connection
	 :dataset ::aggregation/dataset
	 :query ::aggregation.pivot/query)
  :ret any?)

(s/def ::aggregation.bar/bucketColumn (s/nilable ::db.dsv.column.s/columnName))
(s/def ::aggregation.bar/subBucketColumn (s/nilable ::db.dsv.column.s/columnName))
(s/def ::aggregation.bar/metricColumnY (s/nilable ::db.dsv.column.s/columnName))
(s/def ::metricAggregation #{"mean"
                             "sum"
                             "min"
                             "max"
                             "count"
                             "median"
                             "distinct"
                             "q1"
                             "q3"})
(s/def ::aggregation.bar/metricAggregation (s/nilable ::metricAggregation))
(s/def ::aggregation.bar/sort (s/nilable #{"asc" "dsc"}))
(s/def ::aggregation.bar/truncateSize (s/nilable string?))

(s/def ::aggregation.bar/query (s/keys :req-un [::postgres.filter/filters
                                                ::aggregation.bar/bucketColumn]
                                       :opt-un [::aggregation.bar/subBucketColumn
                                                ::aggregation.bar/metricColumnY
                                                ::aggregation.bar/sort
                                                ::aggregation.bar/truncateSize]))

(s/fdef aggregation.bar/query
  :args (s/cat
         :db-conn ::db.s/tenant-connection
	 :dataset ::aggregation/dataset
	 :query ::aggregation.bar/query)
  :ret any?)

(s/def ::aggregation.bubble/bucketColumn ::db.dsv.column.s/columnName)
(s/def ::aggregation.bubble/metricColumn (s/nilable ::db.dsv.column.s/columnName))
(s/def ::aggregation.bubble/metricAggregation (s/nilable ::metricAggregation))

(s/def ::aggregation.bubble/query (s/keys :req-un [::postgres.filter/filters
                                                   ::aggregation.bubble/bucketColumn
                                                   ::aggregation.bubble/metricColumn
                                                   ::aggregation.bubble/metricAggregation]))

(s/fdef aggregation.bubble/query
  :args (s/cat
         :db-conn ::db.s/tenant-connection
	 :dataset ::aggregation/dataset
	 :query ::aggregation.bubble/query)
  :ret any?)

(s/def ::aggregation.line/metricColumnX ::db.dsv.column.s/columnName)
(s/def ::aggregation.line/metricColumnY ::db.dsv.column.s/columnName)
(s/def ::aggregation.line/metricAggregation (s/nilable ::metricAggregation))

(s/def ::aggregation.line/query (s/keys :req-un [::postgres.filter/filters
                                                 ::aggregation.line/metricColumnX
                                                 ::aggregation.line/metricColumnY]
                                        :opt-un [::aggregation.line/metricAggregation]))

(s/fdef aggregation.line/query
  :args (s/cat
         :db-conn ::db.s/tenant-connection
	 :dataset ::aggregation/dataset
	 :query ::aggregation.line/query)
  :ret any?)

(s/def ::aggregation.scatter/metricColumnX ::db.dsv.column.s/columnName)
(s/def ::aggregation.scatter/metricColumnY ::db.dsv.column.s/columnName)
(s/def ::aggregation.scatter/metricColumnSize (s/nilable ::db.dsv.column.s/columnName))
(s/def ::aggregation.scatter/bucketColumnCategory (s/nilable ::db.dsv.column.s/columnName))
(s/def ::aggregation.scatter/datapointLabelColumn (s/nilable ::db.dsv.column.s/columnName))
(s/def ::aggregation.scatter/bucketColumn (s/nilable ::db.dsv.column.s/columnName))
(s/def ::aggregation.scatter/metricAggregation (s/nilable ::metricAggregation))

(s/def ::aggregation.scatter/query (s/keys :req-un [::postgres.filter/filters
                                                    ::aggregation.scatter/metricAggregation
                                                    ::aggregation.scatter/metricColumnX
                                                    ::aggregation.scatter/metricColumnY]
                                           :opt-un [::aggregation.scatter/bucketColumn
                                                    ::aggregation.scatter/metricColumnSize
                                                    ::aggregation.scatter/datapointLabelColumn
                                                    ::aggregation.scatter/bucketColumnCategory]))

(s/fdef aggregation.scatter/query
  :args (s/cat
         :db-conn ::db.s/tenant-connection
	 :dataset ::aggregation/dataset
	 :query ::aggregation.scatter/query)
  :ret any?)

File: client/src/utilities/aggregation.js
import { isEqual } from 'lodash';
import { checkUndefined } from './utils';

export const specIsValidForApi = (spec, vType) => {
  let anyLayerInvalid;

  switch (vType) {
    case 'scatter':
      if (spec.metricColumnX === null || spec.metricColumnY === null) {
        return false;
      }
      break;
    case 'bar':
      if (spec.bucketColumn === null) {
        return false;
      }
      break;
    case 'bubble':
      if (spec.bucketColumn === null) {
        return false;
      }
      break;
    case 'line':
    case 'area':
      if (spec.metricColumnX === null || spec.metricColumnY === null) {
        return false;
      }
      break;
    case 'pivot table':
      if (spec.aggregation !== 'count' && spec.valueColumn == null) {
        return false;
      }
      break;
    case 'pie':
    case 'donut':
    case 'polararea':
      if (spec.bucketColumn === null) {
        return false;
      }
      break;
    case 'map':
      if (spec.layers.length === 0) {
        return false;
      }
      anyLayerInvalid = spec.layers.some(
        // Function that should return true if the current layer is invalid
        (layer) => {
          const missingDatasetId = !layer.datasetId;
          const missingRasterId = !layer.rasterId;

          if (layer.layerType === 'geo-shape') {
            return missingDatasetId || !layer.geom;
          } else if (layer.layerType === 'raster') {
            return missingRasterId;
          }

          // Some old maps don't have a layer-type in the layer spec, so just assume it's geopoint
          const missingGeom = !layer.geom;
          const missingLatLong = !(layer.latitude && layer.longitude);
          const missingAnyValidLocation = (missingGeom && missingLatLong);

          return (missingDatasetId || missingAnyValidLocation);
        }
      );

      if (anyLayerInvalid) {
        return false;
      }
      break;
    default:
      return false;
  }
  return true;
};

export const getNeedNewAggregation = (
  newV = { spec: {} },
  oldV = { spec: {} },
  optionalVizType
) => {
  const vType = newV.visualisationType || optionalVizType;

  if ((newV && !oldV) || (vType !== (oldV.visualisationType || 'map'))) {
    console.log('getNeedNewAggregation', newV, oldV, vType, oldV.visualisationType);
    return true;
  }
  const newLegendOrder = checkUndefined(newV, 'spec', 'legend', 'order') || {};
  const oldLegendOrder = checkUndefined(oldV, 'spec', 'legend', 'order') || {};

  switch (vType) {
    case 'bar':
      return Boolean(
        newV.datasetId !== oldV.datasetId ||
        newV.spec.metricColumnX !== oldV.spec.metricColumnX ||
          newV.spec.metricColumnY !== oldV.spec.metricColumnY ||
          !isEqual(newV.spec.metricColumnsY, oldV.spec.metricColumnsY) ||
        newV.spec.metricAggregation !== oldV.spec.metricAggregation ||
        newV.spec.subBucketMethod !== oldV.spec.subBucketMethod ||
        newV.spec.bucketColumn !== oldV.spec.bucketColumn ||
        newV.spec.subBucketColumn !== oldV.spec.subBucketColumn ||
        newV.spec.sort !== oldV.spec.sort ||
        newV.spec.truncateSize !== oldV.spec.truncateSize ||
        !isEqual(newV.spec.filters, oldV.spec.filters)
      );
    case 'bubble':
      return Boolean(
        newV.datasetId !== oldV.datasetId ||
        newV.spec.metricColumn !== oldV.spec.metricColumn ||
        newV.spec.bucketColumn !== oldV.spec.bucketColumn ||
        newV.spec.metricAggregation !== oldV.spec.metricAggregation ||
        !isEqual(newV.spec.filters, oldV.spec.filters)
      );
    case 'pivot table':
      return Boolean(
        newV.datasetId !== oldV.datasetId ||
        newV.spec.aggregation !== oldV.spec.aggregation ||
        newV.spec.valueColumn !== oldV.spec.valueColumn ||
        newV.spec.categoryColumn !== oldV.spec.categoryColumn ||
        newV.spec.rowColumn !== oldV.spec.rowColumn ||
        !isEqual(newV.spec.filters, oldV.spec.filters)
      );
    case 'pie':
    case 'donut':
    case 'polararea':
      return Boolean(
        newV.datasetId !== oldV.datasetId ||
        newV.spec.bucketColumn !== oldV.spec.bucketColumn ||
        newV.spec.sort !== oldV.spec.sort ||
        !isEqual(newV.spec.filters, oldV.spec.filters) ||
        newLegendOrder.mode !== oldLegendOrder.mode ||
        !isEqual(newLegendOrder.list, oldLegendOrder.list)
      );
    case 'line':
    case 'area':
      return Boolean(
        newV.datasetId !== oldV.datasetId ||
        newV.spec.metricColumnX !== oldV.spec.metricColumnX ||
        newV.spec.metricColumnY !== oldV.spec.metricColumnY ||
        newV.spec.metricAggregation !== oldV.spec.metricAggregation ||
        !isEqual(newV.spec.filters, oldV.spec.filters)
      );
    case 'scatter':
      return Boolean(
        newV.datasetId !== oldV.datasetId ||
        newV.spec.metricColumnX !== oldV.spec.metricColumnX ||
        newV.spec.metricColumnY !== oldV.spec.metricColumnY ||
        newV.spec.metricColumnSize !== oldV.spec.metricColumnSize ||
        newV.spec.bucketColumnCategory !== oldV.spec.bucketColumnCategory ||
        newV.spec.metricAggregation !== oldV.spec.metricAggregation ||
        newV.spec.bucketColumn !== oldV.spec.bucketColumn ||
        newV.spec.datapointLabelColumn !== oldV.spec.datapointLabelColumn ||
        !isEqual(newV.spec.filters, oldV.spec.filters)
      );

    case 'map':
      return Boolean(
        newV.datasetId !== oldV.datasetId ||
        newV.rasterId !== oldV.rasterId ||
        newV.aggregationDataset !== oldV.aggregationDataset ||
        newV.aggregationColumn !== oldV.aggregationColumn ||
        newV.aggregationGeomColumn !== oldV.aggregationGeomColumn ||
        newV.aggregationMethod !== oldV.aggregationMethod ||
        newV.shapeLabelColumn !== oldV.shapeLabelColumn ||
        newV.latitude !== oldV.latitude ||
        newV.longitude !== oldV.longitude ||
        newV.geom !== oldV.geom ||
        newV.pointColorColumn !== oldV.pointColorColumn ||
        newV.pointSize !== oldV.pointSize ||
        newV.gradientColor !== oldV.gradientColor ||
        newV.startColor !== oldV.startColor ||
        newV.endColor !== oldV.endColor ||
        !isEqual(newV.filters, oldV.filters) ||
        !isEqual(newV.popup, oldV.popup) ||
        !isEqual(newV.pointColorMapping, oldV.pointColorMapping)
      );
    default:
      throw new Error(`Unknown visualisation type ${vType} supplied to getNeedNewAggregation`);
  }
};
